src/message/kafka/index.ts

```diff
@@ -0,0 +1,248 @@
+
+/**
+ * Bindings for Kafka transport
+ * @implements {@linkcode Binding}
+ */
+ const Kafka: Binding = {
+  binary: toBinaryKafkaMessage,
+  structured: toStructuredKafkaMessage,
+  toEvent: deserializeKafkaMessage,
+  isEvent: isKafkaEvent,
+};
+
+/**
+ * Extends the base Message type to include
+ * Kafka-specific fields
+ */
+interface KafkaMessage<T> extends Message<T> {
+  key: string | Buffer
+  value: T | string | Buffer | unknown
+  timestamp?: string
+}
+
+// Export the binding implementation and message interface
+export {
+  Kafka,
+  KafkaMessage
+};
+
+/**
+ * Serialize a CloudEvent for Kafka in binary mode
+ * @implements {Serializer}
+ * @see https://github.com/cloudevents/spec/blob/v1.0.1/kafka-protocol-binding.md#32-binary-content-mode
+ *
+ * @param {CloudEvent<T>} event The event to serialize
+ * @returns {KafkaMessage<T>} a KafkaMessage instance
+ */
+function toBinaryKafkaMessage<T>(event: CloudEvent<T>): KafkaMessage<T> {
+  // 3.2.1. Content Type
+  // For the binary mode, the header content-type property MUST be mapped directly
+  // to the CloudEvents datacontenttype attribute.
+  const headers: Headers = { 
+    ...{ [CONSTANTS.HEADER_CONTENT_TYPE]: event.datacontenttype }, 
+    ...headersFor(event) };
+  return {
+    headers,
+    key: createKey(event),
+    value: event.data,
+    body: event.data,
+    timestamp: timestamp(event.time)
+  };
+}
+
+/**
+ * Serialize a CloudEvent for Kafka in structured mode
+ * @implements {Serializer}
+ * @see https://github.com/cloudevents/spec/blob/v1.0.1/kafka-protocol-binding.md#33-structured-content-mode
+ *
+ * @param {CloudEvent<T>} event the CloudEvent to be serialized
+ * @returns {KafkaMessage<T>} a KafkaMessage instance
+ */
+ function toStructuredKafkaMessage<T>(event: CloudEvent<T>): KafkaMessage<T> {
+  if (event.data_base64) {
+    // The event's data is binary - delete it
+    event = event.cloneWith({ data: undefined });
+  }
+  const value = event.toString();
+  return {
+    key: createKey(event),
+    value,
+    headers: {
+      [CONSTANTS.HEADER_CONTENT_TYPE]: CONSTANTS.DEFAULT_CE_CONTENT_TYPE,
+    },
+    body: value,
+    timestamp: timestamp(event.time)
+  };
+}
+
+/**
+ * Converts a Message to a CloudEvent
+ * @implements {Deserializer}
+ *
+ * @param {Message} message the incoming message
+ * @return {CloudEvent} A new {CloudEvent} instance
+ */
+function deserializeKafkaMessage<T>(message: Message<T>): CloudEvent<T> | CloudEvent<T>[] {
+  if (!isKafkaEvent(message)) {
+    throw new ValidationError("No CloudEvent detected");
+  }
+  if (!(message as KafkaMessage<T>).value) {
+    throw new ValidationError("Value is null or undefined");
+  }
+  if (!message.headers) {
+    throw new ValidationError("Headers are null or undefined");
+  }
+  const cleanHeaders: Headers = sanitize(message.headers);
+  const mode: Mode = getMode(cleanHeaders);
+  switch (mode) {
+    case Mode.BINARY:
+      return parseBinary(message as KafkaMessage<T>);
```

Comment:
Type assertion `message as KafkaMessage<T>` can be done once (or fewer times) above.

Reply:
I tried to simplify this, but didn't have much luck at an elegant solution. If you can suggest a specific change for this that compiles I'd be happy to use it.